// @trycourier/courier/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CourierClient = void 0;
const environments = __importStar(require("./environments"));
const core = __importStar(require("./core"));
const url_join_1 = __importDefault(require("url-join"));
const errors = __importStar(require("./errors/index"));
const Client_1 = require("./api/resources/audiences/client/Client");
const Client_2 = require("./api/resources/auditEvents/client/Client");
const Client_3 = require("./api/resources/authTokens/client/Client");
const Client_4 = require("./api/resources/automations/client/Client");
const Client_5 = require("./api/resources/brands/client/Client");
const Client_6 = require("./api/resources/bulk/client/Client");
const Client_7 = require("./api/resources/lists/client/Client");
const Client_8 = require("./api/resources/messages/client/Client");
const Client_9 = require("./api/resources/notifications/client/Client");
const Client_10 = require("./api/resources/profiles/client/Client");
const Client_11 = require("./api/resources/templates/client/Client");
const Client_12 = require("./api/resources/tenants/client/Client");
const Client_13 = require("./api/resources/translations/client/Client");
const Client_14 = require("./api/resources/users/client/Client");
class CourierClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * Use the send API to send a message to one or more recipients.
     *
     * @param {Courier.SendMessageRequest} request
     * @param {CourierClient.IdempotentRequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await courier.send({
     *         message: {
     *             content: {},
     *             data: {
     *                 "string": {
     *                     "key": "value"
     *                 }
     *             },
     *             brand_id: "string",
     *             channels: {
     *                 "string": {}
     *             },
     *             context: {},
     *             metadata: {},
     *             providers: {
     *                 "string": {}
     *             },
     *             routing: {},
     *             timeout: {},
     *             delay: {},
     *             expiry: {},
     *             to: {}
     *         }
     *     })
     */
    send(request, requestOptions) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const _response = yield ((_a = this._options.fetcher) !== null && _a !== void 0 ? _a : core.fetcher)({
                url: (0, url_join_1.default)((_b = (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.CourierEnvironment.Production, "/send"),
                method: "POST",
                headers: {
                    Authorization: yield this._getAuthorizationHeader(),
                    "X-Fern-Language": "JavaScript",
                    "X-Fern-SDK-Name": "@trycourier/courier",
                    "X-Fern-SDK-Version": "v6.1.2",
                    "X-Fern-Runtime": core.RUNTIME.type,
                    "X-Fern-Runtime-Version": core.RUNTIME.version,
                    "Idempotency-Key": (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.idempotencyKey) != null ? requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.idempotencyKey : undefined,
                    "X-Idempotency-Expiration": (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.idempotencyExpiry) != null ? requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.idempotencyExpiry : undefined,
                },
                contentType: "application/json",
                body: request,
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
            });
            if (_response.ok) {
                return _response.body;
            }
            if (_response.error.reason === "status-code") {
                throw new errors.CourierError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                });
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.CourierError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                    });
                case "timeout":
                    throw new errors.CourierTimeoutError();
                case "unknown":
                    throw new errors.CourierError({
                        message: _response.error.errorMessage,
                    });
            }
        });
    }
    get audiences() {
        var _a;
        return ((_a = this._audiences) !== null && _a !== void 0 ? _a : (this._audiences = new Client_1.Audiences(this._options)));
    }
    get auditEvents() {
        var _a;
        return ((_a = this._auditEvents) !== null && _a !== void 0 ? _a : (this._auditEvents = new Client_2.AuditEvents(this._options)));
    }
    get authTokens() {
        var _a;
        return ((_a = this._authTokens) !== null && _a !== void 0 ? _a : (this._authTokens = new Client_3.AuthTokens(this._options)));
    }
    get automations() {
        var _a;
        return ((_a = this._automations) !== null && _a !== void 0 ? _a : (this._automations = new Client_4.Automations(this._options)));
    }
    get brands() {
        var _a;
        return ((_a = this._brands) !== null && _a !== void 0 ? _a : (this._brands = new Client_5.Brands(this._options)));
    }
    get bulk() {
        var _a;
        return ((_a = this._bulk) !== null && _a !== void 0 ? _a : (this._bulk = new Client_6.Bulk(this._options)));
    }
    get lists() {
        var _a;
        return ((_a = this._lists) !== null && _a !== void 0 ? _a : (this._lists = new Client_7.Lists(this._options)));
    }
    get messages() {
        var _a;
        return ((_a = this._messages) !== null && _a !== void 0 ? _a : (this._messages = new Client_8.Messages(this._options)));
    }
    get notifications() {
        var _a;
        return ((_a = this._notifications) !== null && _a !== void 0 ? _a : (this._notifications = new Client_9.Notifications(this._options)));
    }
    get profiles() {
        var _a;
        return ((_a = this._profiles) !== null && _a !== void 0 ? _a : (this._profiles = new Client_10.Profiles(this._options)));
    }
    get templates() {
        var _a;
        return ((_a = this._templates) !== null && _a !== void 0 ? _a : (this._templates = new Client_11.Templates(this._options)));
    }
    get tenants() {
        var _a;
        return ((_a = this._tenants) !== null && _a !== void 0 ? _a : (this._tenants = new Client_12.Tenants(this._options)));
    }
    get translations() {
        var _a;
        return ((_a = this._translations) !== null && _a !== void 0 ? _a : (this._translations = new Client_13.Translations(this._options)));
    }
    get users() {
        var _a;
        return ((_a = this._users) !== null && _a !== void 0 ? _a : (this._users = new Client_14.Users(this._options)));
    }
    _getAuthorizationHeader() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const bearer = (_a = (yield core.Supplier.get(this._options.authorizationToken))) !== null && _a !== void 0 ? _a : process === null || process === void 0 ? void 0 : process.env["COURIER_AUTH_TOKEN"];
            if (bearer == null) {
                throw new errors.CourierError({
                    message: "Please specify COURIER_AUTH_TOKEN when instantiating the client.",
                });
            }
            return `Bearer ${bearer}`;
        });
    }
}
exports.CourierClient = CourierClient;
