// @trycourier/courier/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.translations = exports.users = exports.tenants = exports.templates = exports.send = exports.profiles = exports.notifications = exports.messages = exports.lists = exports.bulk = exports.commons = exports.brands = exports.automations = exports.authTokens = exports.auditEvents = exports.audiences = void 0;
exports.audiences = __importStar(require("./audiences"));
__exportStar(require("./audiences/types"), exports);
exports.auditEvents = __importStar(require("./auditEvents"));
__exportStar(require("./auditEvents/types"), exports);
exports.authTokens = __importStar(require("./authTokens"));
__exportStar(require("./authTokens/types"), exports);
exports.automations = __importStar(require("./automations"));
__exportStar(require("./automations/types"), exports);
exports.brands = __importStar(require("./brands"));
__exportStar(require("./brands/types"), exports);
exports.commons = __importStar(require("./commons"));
__exportStar(require("./commons/types"), exports);
exports.bulk = __importStar(require("./bulk"));
__exportStar(require("./bulk/types"), exports);
exports.lists = __importStar(require("./lists"));
__exportStar(require("./lists/types"), exports);
exports.messages = __importStar(require("./messages"));
__exportStar(require("./messages/types"), exports);
exports.notifications = __importStar(require("./notifications"));
__exportStar(require("./notifications/types"), exports);
exports.profiles = __importStar(require("./profiles"));
__exportStar(require("./profiles/types"), exports);
exports.send = __importStar(require("./send"));
__exportStar(require("./send/types"), exports);
exports.templates = __importStar(require("./templates"));
__exportStar(require("./templates/types"), exports);
exports.tenants = __importStar(require("./tenants"));
__exportStar(require("./tenants/types"), exports);
exports.users = __importStar(require("./users"));
__exportStar(require("./commons/errors"), exports);
exports.translations = __importStar(require("./translations"));
__exportStar(require("./audiences/client/requests"), exports);
__exportStar(require("./auditEvents/client/requests"), exports);
__exportStar(require("./authTokens/client/requests"), exports);
__exportStar(require("./brands/client/requests"), exports);
__exportStar(require("./bulk/client/requests"), exports);
__exportStar(require("./lists/client/requests"), exports);
__exportStar(require("./messages/client/requests"), exports);
__exportStar(require("./notifications/client/requests"), exports);
__exportStar(require("./profiles/client/requests"), exports);
__exportStar(require("./templates/client/requests"), exports);
__exportStar(require("./tenants/client/requests"), exports);
